### ISSUE TITLE
Build system

### ISSUE THREAD

**Issue by bryanperris** [bryanperris/winN64dev#1]
Prefer to use a configured build system like make or cmake to compile the source and build the rom without needing to write python scripts.

**Comment by cxd4** [comment]
https://github.com/bryanperris/winN64dev/blob/bcf2f2b957184defe823037c2ea25d67ea34a97d/build.py
I was wondering what about perhaps a shell script (or a Windows-only command shell batch file I guess since it appears this tool chain may be Windows-only)?  Unfamiliar with Python scripting but I can see the gist of these functions being doable in a batch file of some type.

Or is it preferred to have Makefile or CMake set up for more dynamic features?

**Comment by bdragoncore** [comment]
I found it easier to write it in python than the old windows shell script language. I would think that using a real build system would be better with dealing with the gcc environment. Anyone writing c code would be more comfortable with cmake or make than a hacked python script. I don't know much how to reconfigure a build system to work with a specific bin directory and specific commands it must call after emitting the elf file.

**Comment by bdragoncore** [comment]
Also the build system needs to add the cygwin dlls to the path environment, so the gcc binaries can execute.

**Comment by bdragoncore** [comment]
Cxd4 if you are interested in making make work with the cygwin compiled gnu toolchain or providing any other scripts I would grant you access to the repo. I won't be doing much with this until I finish my n64 emulator. I used what I had here to build a dummy test rom to analyze how  basic homebrew roms are composed. Looks like all homebrew roms piggyback the generic boot logic but provide an accurate rom checksum in the header. Still much to learn though.

**Comment by cxd4** [comment]
Alright 1st sorry about delay in reply as that was just a 10-day absence of a working PC (boot issues).

I am familiar with native GCC for GNU or *nix as well as MinGW for Windows (which is free of the full POSIX emulation dependency that Cygwin for Windows would require, so no issue there), but I am not really able to deal with Makefile syntax.  No help there I'm afraid.  (CMake is probably more readable, though does not work out of the box on e.g. Win32 without installing additional tools.)  Although, I do like that Unix shell scipts will work both on nix as well as Windows; similarly a Windows batch file can also run under Linux out of the box on some distributions.  I could write either one.

I guess the real issue I was meaning to pose here, is just that right now (with or without the Makefile approach), I do not have Windows running here or knowledge of how to use this toolchain outside of Windows.  Then I would be able to test it, and write a shell script that can operate and do the N64 build on either platform.  (Windows can execute shell scripts written for Linux when they are run through Git shell.)

**Comment by bdragoncore** [comment]
I see, I do think unix shell script wrapped in a windows batch script would be good, so we can include cygwin bash.exe.  The windows shell script would need to setup the cygwin environment vars so DLLs are found by bash itself.  Then the rest can be handled by the unix scripts, but it seems too complex to use shell scripts to make GCC fly, when maybe I should play with CodeBlocks IDE to have it take care of setting up GCC along with some preconfigured GCC arguments and custom build events.  Since I intended to support N64 Rom building tools for the Windows side, most likely Windows developers will rather use CodeBlocks compared to command line anyways.  It would be possible to get VS2015 to use GCC instead, but that might include a million years into building the VS extensions to make VS use a custom C/C++ toolchain rather than MSbuild etc.

https://github.com/DragonMinded/libdragon

Based on that documentation, setting up the toolchain is pretty simple under any Linux machine, cygwin was much tricker.  I forget if I tried using MinGW, but I know that MinGW 2 makes the easier to manage packages and such.

**Comment by cxd4** [comment]
Code::Blocks sounds fine as well.  I actually had not gotten around to upgrading to that (was long ago using its dead predecessor Dev-Cpp for an IDE to MinGW on Windows), but I'm able to follow it better than the syntax used in most Makefile implementations.  I've typically practiced command scripts mostly because they're lighter and don't involve frameworks, IDEs or other dependencies (that and a simple double-click is quick and convenient enough, whether done on Windows or not).

Hmm although I gave libdragon a try from their README directions and am getting an eventual

```
checking dynamic linker characteristics... configure: error: Link tests are not allowed after GCC_NO_EXECUTABLES.
make[1]: *** [configure-zlib] Error 1
```

so guess I have not yet figured out beginning with libdragon yet.  It's worth noting that I am totally new to actually doing N64 homebrew despite having a bigger interest in it, so that could be anything.

**Comment by bdragoncore** [comment]
I would Google what that env variable is used for. I am the same boat with you, I never written any homebrew and but I wanted to compile basic routines to debug and analyze in my own emulator.

**Comment by bdragoncore** [comment]
https://gcc.gnu.org/ml/gcc-help/2012-07/msg00018.html

You might be missing a library compiled for mips that gcc expects for the mips gcc. Personally never seen that kind of error before.

**Comment by cxd4** [comment]
Yes, that's more than possible.  Formerly I had to manually fix my operating system due to an asynchronous package upgrade which caused my version of glibc to not correspond to the locale of the rest of my system.  I have not yet looked in detail at getting libdragon to work, but otherwise I would be happy to look at fixing/amending my install of glibc here.

Though, with libdragon I do like what I'm seeing.  The other, more standard options (I think one of them I am reading about is Psy-Q) do not really seem to respect the standard code routines in the C language and make you use custom APIs for a lot more things, but libdragon looks more portable, not to mention available freely.

Psy-Q I'm taking it seems to be more of a Windows-specific thing.  This repository seems to be using something more freely available I must have forgotten about from years back, but I think more important to me even than being available outside of Windows, is integrating with the standard C library instead of using external, new APIs like official N64 code seems to.

**Comment by bdragoncore** [comment]
When I researched online for n64 brewing, I mostly found broken links, but after finding libdragon, I saw it includes everything you need to get started. I was excited after finding this and took the time to get it run on Windows. So I left this repo here for anyone interested in expanding the build scripts or making a free ide work with it. 

BTW, I noticed many rcp projects on your github. Someday I want to attempt porting rcp lle code to fsharp and execute the logic with the fsharp to opencl toolkit to perform dsp and graphics routines on the host gpu. Also non opencl mode would be supported.

**Comment by cxd4** [comment]
The next time I boot into a Windows partition I will give it a try with the version here in this repo.  It's probably better that I tackle it that way so that I can have a comparison and test by trial and error on how to make a shell script, VS2010 project or GNU Make file of some kind.  Ordinarily I would say translate the existing make code, but unfortunately I do not come from a background of knowing Python.

Also not included in my background is C# history, but it's nice to see you have a project for that with emulation.  I may not be fond of C# particularly, but I think you can never really go wrong with the exercise of re-expressing knowledge bases in different languages.  I wanted to learn OpenCL once to see if it had any pertinence with RSP optimizations but found out it was something different; instead I just use OpenAL and OpenGL.

**Comment by bdragoncore** [comment]
I added you to the project team anyways, so you can push anything.  Perhaps someday, we can collaborate on RSP/RDP emulation in C#/F#, to get down the basic structures, algorithms, IO API, etc.  I can still use https://sharpgl.codeplex.com/ for standard OGL graphics, and also support the HLE version of the RCP.
